web/src/app/shared/store/classification-store/classification.state.ts

```diff
@@ -1,46 +1,187 @@
+
+  @Action(SelectClassification)
+  selectClassification(ctx: StateContext<ClassificationStateModel>, action: SelectClassification): Observable<any|null> {
+    if (typeof action.classificationId !== 'undefined') {
+      return this.classificationsService.getClassification(action.classificationId).pipe(take(1), tap(
+        selectedClassification => {
+          ctx.patchState({
+            selectedClassification,
+            action: null
+          });
+        }
+      ));
+    }
+    return of(null);
+  }
+
+  @Action(DeselectClassification)
+  deselectClassification(ctx: StateContext<ClassificationStateModel>): Observable<null> {
+    ctx.patchState({
+      selectedClassification: undefined,
+      action: null
+    });
+    return of(null);
   }
 
   @Action(InitializeStore)
   initializeStore(ctx: StateContext<ClassificationStateModel>): Observable<any> {
     return this.categoryService.getClassificationCategoriesByType().pipe(
-      tap(classificationTypes => {
+      take(1), tap(classificationTypes => {
         ctx.setState({
           ...ctx.getState(),
           classificationTypes,
+          classifications: undefined,
           selectedClassificationType: Object.keys(classificationTypes)[0],
         });
       }),
     );
   }
 
+  @Action(GetClassifications)
+  getClassifications(ctx: StateContext<ClassificationStateModel>): Observable<any> {
+    const { selectedClassificationType } = ctx.getState();
+    return this.classificationsService.getClassifications(selectedClassificationType).pipe(
+      take(1), tap(classifications => {
+        classifications.forEach(classification => {
+          classification.children = !classification.children ? [] : classification.children;
+        });
+        ctx.patchState({
+          classifications: [...classifications]
+        });
+      }),
+    );
+  }
+
+  @Action(CreateClassification)
+  createClassification(ctx: StateContext<ClassificationStateModel>, action: CreateClassification): Observable<any> {
+    return this.classificationsService.postClassification(action.classification).pipe(
+      take(1), tap(classification => {
+        ctx.patchState(
+          {
+            classifications: [...ctx.getState().classifications, classification],
+            selectedClassification: classification,
+            action: null
+          }
+        );
+      })
+    );
+  }
+
+  @Action(SaveClassification)
+  saveClassification(ctx: StateContext<ClassificationStateModel>, action: SaveClassification): Observable<any> {
+    return this.classificationsService.putClassification(action.classification).pipe(
+      take(1), tap(savedClassification => {
+        ctx.patchState({
+          classifications: ctx.getState().classifications.map(currentClassification => {
+            if (currentClassification.classificationId === savedClassification.classificationId) {
+              return savedClassification;
+            }
+            return currentClassification;
+          }),
```

Comment:
save this map function into a variable within this action so that the patchState does not look complicated?